File: 278.第一个错误的版本.py
#
# @lc app=leetcode.cn id=278 lang=python
#
# [278] 第一个错误的版本
#

# @lc code=start
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        l, r = 1, n
        while  l <= r:
            m = (l+r) // 2
            if isBadVersion(m) > m * isBadVersion(m - 1):
                return m
            elif isBadVersion(m):
                r = m - 1
            else:
                l = m + 1

        
# @lc code=end

File: 278.第一个错误的版本.cpp
/*
 * @lc app=leetcode.cn id=278 lang=cpp
 *
 * [278] 第一个错误的版本
 */

// @lc code=start
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        long int l = 0, r = n;
        while(l <= r){
            int mid = (l +r) / 2;
            if(isBadVersion(mid) == true)
            {
                if(isBadVersion(mid-1) == false)
                    return mid;
                r = mid - 1;

            }
            else{
                l = mid + 1;
            }
        }
        return 0;
        
    }
};
// @lc code=end

